source/qb64pe.bas

```diff
@@ -3413,6 +3418,88 @@ DO
+            EmbedFile$ = _TRIM$(MID$(a3$, bra + 1, ket - bra - 1))
+            IF LEN(EmbedFile$) = 0 GOTO errmes
+            'check for handle
+            bra = INSTR(ket + 1, a3u$, "'"): IF bra = 0 GOTO errmes
+            ket = INSTR(bra + 1, a3u$, "'"): IF ket = 0 GOTO errmes
+            EmbedHandle$ = _TRIM$(MID$(a3$, bra + 1, ket - bra - 1))
+            IF LEN(EmbedHandle$) = 0 GOTO errmes
+            'fix layout
+            layout$ = SCase$("$Embed:'") + EmbedFile$ + "','" + EmbedHandle$ + "'" + MID$(a3$, ket + 1)
+            'verify path/file existence
+            EmbedPath$ = ""
+            IF LEFT$(EmbedFile$, 2) = "./" OR LEFT$(EmbedFile$, 2) = ".\" THEN
+                IF NoIDEMode THEN
+                    EmbedPath$ = path.source$
+                    IF LEN(EmbedPath$) > 0 AND RIGHT$(EmbedPath$, 1) <> pathsep$ THEN EmbedPath$ = EmbedPath$ + pathsep$
+                ELSE
+                    IF LEN(ideprogname) THEN EmbedPath$ = idepath$ + pathsep$
+                END IF
+                EmbedFile$ = EmbedPath$ + MID$(EmbedFile$, 3)
+            ELSEIF INSTR(EmbedFile$, "/") OR INSTR(EmbedFile$, "\") THEN
+                FOR i = LEN(EmbedFile$) TO 1 STEP -1
+                    IF MID$(EmbedFile$, i, 1) = "/" OR MID$(EmbedFile$, i, 1) = "\" THEN
+                        EmbedPath$ = LEFT$(EmbedFile$, i)
+                        EmbedFileOnly$ = MID$(EmbedFile$, i + 1)
+                        IF _DIREXISTS(EmbedPath$) = 0 THEN a$ = "File '" + EmbedFileOnly$ + "' not found": GOTO errmes
+                        currentdir$ = _CWD$
+                        CHDIR EmbedPath$
+                        EmbedPath$ = _CWD$
+                        CHDIR currentdir$
+                        EmbedFile$ = EmbedPath$ + pathsep$ + EmbedFileOnly$
+                        EXIT FOR
+                    END IF
+                NEXT
+            END IF
+            IF _FILEEXISTS(EmbedFile$) = 0 THEN a$ = "File '" + EmbedFile$ + "' not found": GOTO errmes
+            'verify handle validity (A-Z/a-z/0-9, begin with letter)
+            i = ASC(EmbedHandle$, 1)
+            IF (i < ASC("A")) OR (i > ASC("Z") AND i < ASC("a")) OR (i > ASC("z")) THEN
+                a$ = "First char of Embed-Handle '" + EmbedHandle$ + "' must be a letter": GOTO errmes
+            END IF
+            FOR j = 2 TO LEN(EmbedHandle$)
+                i = ASC(EmbedHandle$, j)
+                IF (i < ASC("0")) OR (i > ASC("9") AND i < ASC("A")) OR (i > ASC("Z") AND i < ASC("a")) OR (i > ASC("z")) THEN
+                    a$ = "Embed-Handle '" + EmbedHandle$ + "' has invalid chars, use A-Z/a-z/0-9 only": GOTO errmes
+                END IF
+            NEXT j
+            'check for duplicate definitions
+            eflUB = UBOUND(embedFileList$, 2)
+            FOR i = 0 TO eflUB
+                IF embedFileList$(eflFile, i) = EmbedFile$ THEN
+                    a$ = "File '" + EmbedFile$ + "' was already embedded in line"
```

Comment:
Wouldn't you just need one embed, even in that case?

image1 = _LOADIMAGE(_Embedded(foo$), 32, "memory")
image2 = _LOADIMAGE(_Embedded(foo$), 32, "memory")
image3 = _LOADIMAGE(_Embedded(foo$), 32, "memory")

The embed is just the image file itself in memory, and just like if it was on the drive, you'd still need to _LoadImage it before it'd become a valid, working image handle.

Reply:
My thinking is not a "can you" but rather "should you". If you have multiple embeds that serve completely different purposes then you don't want them to share a handle regardless of whether they use the same file - in the future you may change the file for one of the handles but not both, and you wouldn't want to have to then figure out which spots are supposed to use which handle.

IMO It's the same as saying "two `CONST`s shouldn't have the same value, if they do then just use the same `CONST` in both places". It's technically true, you don't need more than one `CONST` with a particular value, but it misses the point that if one `CONST` is named `test_file` and another is named `logo` they might temporarily (or always) have the same value but they don't serve the same purpose and combining them wouldn't be correct.

Reply:
I don't think it's like having two CONSTs with the same value.  I think it's more like putting the same font file inside your folder and then naming it cour(2).ttf, cout(3).ttf, cour(4).ttf....    At the end of the day, does it matter which of those files you use to load into a font handle?  They're all the exact same font, and unless you have some plan to overwrite those files, there's absolutely no reason to have more than that one.  

I honestly don't think there's any way to overwrite (and save) embedded data inside a program, with that program itself.  Windows doesn't allow that type of virus-like behavior at all.

I'd think a better way to think of it would be like with DATA statements.   Would you ever include the same 1000 lines of data, just to place them under a separate label?   Or would you just make multiple labels for the same data?

label_1:
label_2:
label_3:

DATA "dog", "frog", "fish", "wish", "chicken"
...999 more lines of data

Then, if you really need to, you can always RESTORE label_1, RESTORE label_2, or RESTORE label_3, but you still wouldn't need to have that same 1000 lines repeated again and again in the same program.

At the most, I think allowing multiple labels for the same embed *might* not be so terrible.   $EMBED:'foo.ext', 'handle1', 'handle2', 'handle3'    But the idea of including the same file multiple times, just to have it listed under a separate label to read back from, seems like a bloat of exe size and a waste of an user's memory.

Reply:
Remember, the handle is only in use really with _EMBEDDED, which will decompress the data and then assign it to wherever you want.  For what you're talking about, all you'd need is:

a$ = _EMBEDDED$(foo$)
b$ = _EMBEDDED$(foo$)

Use a$ all you want in one spot.  Use b$ all you want in a completely different spot.  foo$ itself is embedded and wouldn't ever really change on you.

Reply:
I think we're actually in agreement :P On the backend I agree, we should only keep one copy of the file embedded and serve it up for all the handles that use that file, but that's also an optimization we could do later-on if Rho doesn't want to do it here.

If we error on using the same file though, as it currently works, then introducing that optimization won't help people who already duplicated their files to get around the error.